service_config.md

```diff
@@ -0,0 +1,194 @@
+#  Services configuration / k8s operators
+
+## Goal
+
+Describe an approach that can be adopted across the operators to improve the
+deployment aspect and the way secrets and config files are generated. The
+solution proposed in this document, in the first place, is to give an operator
+the ability to render and inject sensitive snippets within the service config.
+
+### Note:
+
+The document is not supposed to cover the implementation details that might
+converge or diverge according to a given operator, the number of deployed
+services, and other potential requirements that are not the same across the
+board.
+
+
+## Proposed approach
+
+The use cases are concentrated in two main aspects:
+
+* Provide a common interface that can be used to build additional config
+  snippets containing sensitive information using k8s `Secrets` instead of
+  `ConfigMaps`
+
+* Provide a common pattern to build regular service config:
+  * they will be processed via golang templates and mounted in a Pod from a
+    `ConfigMap` (which is what is currently happening)
+  * part of them will be rendered in a `Secret` and avoid exposing sensitive
+    information related to the deployment aspect
+
+
+## Basic deployment template
+
+Usually, the basic deployment config information is rendered in a `ConfigMap`.
+However, the idea is to store sensitive information in a `Secret` that, during
+the bootstrap of the service, will be mounted in the service Pod. For this
```

Comment:
Is it OK if an operator does not use init container to bootstrap a service pod but instead this secret will be mounted to the Pod for the lifetime of the Pod?

Reply:
i dont think its a case of is it ok.
that should be the default after this change is adopted.

we should not require any init containers for configs

Reply:
Then I think we should drop the bootstrap wording from this.

Reply:
Not for config, but there are services with different use cases where can just drop the init container: anyway, I'm ok about dropping the bootstrap wording.

Reply:
init containers are still required in some cases. e.g. there ovs-controller requires an init container to run ovs-ctl commands before the service container starts.
Also if services need to be exposed via multus to networks and the the IP information is required for a config, that would be also a use case for an init container for config. That information is only available at pod start, e.g. OvnEncapIP - https://github.com/openstack-k8s-operators/ovs-operator/pull/75 . if there are services which require setting such information in a config it would be required to do so via an init container

Reply:
there may still be use-cases for them but I'm not sure if this is a good example.
this could be done by the use of a kolla_extended_start script
in the ovs-vswichtd container instead of an init container.

https://github.com/openstack-k8s-operators/tcib/blob/main/container-images/kolla/base/start.sh#L12-L15

infact the entire init.sh that is currently there could likely be moved to use kolla_extended_start since that was the exact usecase it extis to enable.

allowing injection of additional operations to be done on start

there may still be other reasons to use an init container but i think the point I'm trying to advance is when considering what to use them for we should hesitate before using them for config file modification.

Reply:
agree we should prevent using them to create service configs where possible. I just wanted to mention that init containers are not bad and there might be use cases as the one above. if kolla_extended_start is used or some init container I guess is up to the component?

Reply:
either works in both cases the script is intended to be injected into the container not baked in so there is little different.

kolla_extended_start exists because just using podman/docker at least when kolla was first started you could could not run init containers.

podman actually allow you to define pods not just container so they might have the concept of init contaienrs in ther model.

kolla_extended_start would work for both openshift an rhel nodes but functionally they serve the same usecase.

so the service should just use what make the most sense in a given context

Reply:
My original intention with my comment was simply that 

> [...] store sensitive information in a `Secret` that, during
the bootstrap of the service, will be mounted in the service Pod.

this implies for me that the Secrets are only mounted during bootstrap. But in reality we mount them for the lifetime of the Pod.

Reply:
+1, I'll change the wording in the next iteration

Reply:
Done.